xds/internal/xdsclient/creds/tls.go

```diff
@@ -0,0 +1,177 @@
+// a file_watcher provider for mTLS transport security. See gRFC A65.
+func NewTLS(jd json.RawMessage) (credentials.Bundle, error) {
+	cfg := &struct {
+		CertificateFile   string `json:"certificate_file"`
+		CACertificateFile string `json:"ca_certificate_file"`
+		PrivateKeyFile    string `json:"private_key_file"`
+	}{}
+
+	tlsConfig := tls.Config{}
+	if err := json.Unmarshal(jd, cfg); err != nil {
+		return nil, err
+	}
+
+	// We cannot simply always use a file_watcher provider because it behaves
+	// slightly differently from the xDS TLS config. Quoting A65:
+	//
+	// > The only difference between the file-watcher certificate provider
+	// > config and this one is that in the file-watcher certificate provider,
+	// > at least one of the "certificate_file" or "ca_certificate_file" fields
+	// > must be specified, whereas in this configuration, it is acceptable to
+	// > specify neither one.
+	//
+	// We only use a file_watcher provider if either one of them or both are
+	// specified.
+	if cfg.CACertificateFile != "" || cfg.CertificateFile != "" || cfg.PrivateKeyFile != "" {
+		// file_watcher currently ignores BuildOptions, but we set them for good
+		// measure.
+		opts := certprovider.BuildOptions{}
+		if cfg.CACertificateFile != "" {
+			opts.WantRoot = true
+		}
+		if cfg.CertificateFile != "" {
+			opts.WantIdentity = true
+		}
+		provider, err := certprovider.GetProvider("file_watcher", jd, opts)
+		if err != nil {
+			// GetProvider fails if jd is invalid, e.g. if only one of private
+			// key and certificate is specified.
+			return nil, fmt.Errorf("failed to get TLS provider: %w", err)
+		}
+		if cfg.CertificateFile != "" {
+			tlsConfig.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+				// Client cert reloading for mTLS.
+				km, err := provider.KeyMaterial(context.Background())
+				if err != nil {
+					return nil, err
+				}
+				if len(km.Certs) != 1 {
+					return nil, fmt.Errorf("there should be exactly exactly one certificate")
+				}
+				return &km.Certs[0], nil
+			}
+			if cfg.CACertificateFile == "" {
+				// no need for a callback to load the CA. Use the normal mTLS
+				// transport credentials.
+				return &tlsBundle{
+					jd:                   jd,
+					transportCredentials: credentials.NewTLS(&tlsConfig),
+				}, nil
+			}
+		}
+		return &tlsBundle{
+			jd: jd,
+			transportCredentials: &caReloadingClientTLSCreds{
+				baseConfig: &tlsConfig,
+				provider:   provider,
+			},
+		}, nil
+	}
+
+	// None of certificate_file and ca_certificate_file are set.
+	// Use the system-wide root certs.
+	return &tlsBundle{
+		jd:                   jd,
+		transportCredentials: credentials.NewTLS(&tlsConfig),
+	}, nil
+}
+
+func (t *tlsBundle) TransportCredentials() credentials.TransportCredentials {
+	return t.transportCredentials
+}
+
+func (t *tlsBundle) PerRPCCredentials() credentials.PerRPCCredentials {
+	// No per-RPC credentials in A65.
```

Comment:
I couldnt find that language in A65. Would be nice if you link the reference.

Reply:
Maybe could be infered from this excerpt from A65
```
setting the channel creds used to communicate with the xDS server
```

Reply:
I replaced the comment with:

```
	// mTLS provides transport credentials only. There are no per-RPC 
	// credentials.
```

TBH it sounds very obvious to me, and that's probably why A65 doesn't mention per RPC at all.

Reply:
SG!